core/domain/user_services.py

```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
     Raises:
         Exception. If a user with the given gae_id already exists.
     """
-    user_settings = get_user_settings(gae_id, strict=False)
+    user_settings = get_user_settings_by_gae_id(gae_id, strict=False)
     if user_settings is not None:
-        raise Exception('User %s already exists.' % gae_id)
+        raise Exception(
+            'User %s already exists for gae_id %s.'
+            % (user_settings.user_id, gae_id))
 
     user_id = user_models.UserSettingsModel.get_new_id('')
     user_settings = UserSettings(
         user_id, gae_id, email, feconf.ROLE_ID_EXPLORATION_EDITOR,
         preferred_language_codes=[constants.DEFAULT_LANGUAGE_CODE])
+    _save_user_auth_details(UserAuth(user_id, gae_id))
     _save_user_settings(user_settings)
     create_user_contributions(user_id, [], [])
     return user_settings
 
 
+def _save_user_auth_details(user_auth):
+    """Commits a user auth object to the datastore.
+
+    Args:
+        user_auth: UserAuth. The user auth domain object to be saved.
+    """
+    user_auth.validate()
+
+    user_auth_dict = {
+        'gae_id': user_auth.gae_id,
+        'pin': user_auth.pin,
+        'parent_user_id': user_auth.parent_user_id,
+        'deleted': user_auth.deleted
+    }
+
+    # If user auth entry with the given user_id does not exist, create a new
+    # one.
+    user_auth_model = user_models.UserAuthModel.get_by_id(user_auth.user_id)
```

Comment:
Won't this raise an exception? I think ``get_by_id`` only returns ``None`` if you do a non-strict retrieval.

Reply:
Sorry I did not what you meant by non-strict retrieval.
This code does work without raising an exception. The tests have covered them. If the id `user_auth.user_id` does not exists in `UserAuthModel` then it will return None, else the model. This same implementation is there is `user_services._save_user_settings()` function also. Note `user_auth.user_id` itself will never be None here.

Reply:
@BenHenning just a note, this is the model layer, which returns None if the NDB model instance is not found. See docs at https://cloud.google.com/appengine/docs/standard/python/ndb/modelclass#class_methods

Reply:
Ah thanks. I did conflate this with the base model get() function.